packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+ * @param optLabelCache Label measurements
+ * @param svgCache SVG cache
+ */
+export const addRenderedLabels = (
+  optLabelCache: LabelMeasurements,
+  svgCache: Map<string, HTMLElement>,
+): LabelCache => {
+  const labelCache: LabelCache = new Map();
+
+  optLabelCache.forEach((value, key) => {
+    if (value.tag === "EquationData") {
+      labelCache.set(key, {
+        tag: "EquationData",
+        width: value.width,
+        height: value.height,
+        rendered: svgCache.get(key)!,
+        descent: value.descent,
+        ascent: value.ascent,
+      });
+    } else {
+      labelCache.set(key, value);
+    }
+  });
+  return labelCache;
+};
+
+/**
+ * Separate labels into measurements and svgs
+ * @param labelCache
+ */
+export const separateRenderedLabels = (
+  labelCache: LabelCache,
+): { optLabelCache: LabelMeasurements; svgCache: Map<string, HTMLElement> } => {
+  const optLabelCache: LabelMeasurements = new Map<string, LabelData>();
+  const svgCache: Map<string, HTMLElement> = new Map();
+
+  labelCache.forEach((value, key) => {
+    optLabelCache.set(key, removeRenderedLabels(value));
+    if (value.tag === "EquationData") {
+      svgCache.set(key, value.rendered);
+    }
+  });
+
+  return { optLabelCache, svgCache };
+};
+
+export const layoutStateToRenderState = (
+  layoutState: LayoutState,
+  svgCache: Map<string, HTMLElement>,
+) => ({
+  ...layoutState,
+  labelCache: addRenderedLabels(layoutState.labelMeasurements, svgCache),
+});
+
+export const collectAndSeparateLabels = async (
+  shapes: Shape<Num>[],
+): Promise<
+  Result<
+    { optLabelCache: LabelMeasurements; svgCache: Map<string, HTMLElement> },
+    PenroseError
+  >
+> => {
+  const convert = mathjaxInit();
+  const labelCache = await collectLabels(shapes, convert);
+  if (labelCache.isErr()) {
+    return Result.err(labelCache.error);
+  }
+
+  return Result.ok(separateRenderedLabels(labelCache.value));
+};
+
+export const stateToLayoutState = (state: State): LayoutState => {
+  return {
+    variation: state.variation,
+    labelMeasurements: separateRenderedLabels(state.labelCache).optLabelCache,
+    canvas: state.canvas,
+    shapes: state.computeShapes(state.varyingValues),
+  };
+};
+
+/**
+ * A `Map`, but after size grows past `maxLen`, old elements are deleted.
+ */
+export class SizeBoundedMap<K, V> {
+  private readonly _map = new Map<K, V>();
+  private readonly maxLen;
+  private readonly onDelete;
+
+  private _keys: K[] = [];
+
+  constructor(maxLen: number, onDelete?: (key: K, val: V) => void) {
+    this.maxLen = maxLen;
+    this.onDelete = onDelete;
+  }
+
+  get = (key: K): V | undefined => {
+    return this._map.get(key);
+  };
+
+  set = (key: K, value: V): void => {
+    // if we're about to grow past our maximum length, find the oldest key
+    // and remove it from _keys and _map.
+    if (this._map.size === this.maxLen) {
+      const toDeleteKey = this._keys.shift();
```

Comment:
Great thinking to build this `SizeBoundedMap` type! One comment: this usage of `shift` turns `set` from $O(1)$ to $O(n)$; is there a way you can structure it so it remains $O(1)$?

Reply:
Doesn't really matter right now since I see that you have `numWorkersToKeep = 3` and `numStepSequencesToKeep = 3`, but I feel like we've gotten bitten by these sorts of time complexity issues in the past.

Reply:
Good point--javascript should really have a builtin linked list type? I'll implement one real quick.